### PR TITLE
Add a Selector type

### PR DESCRIPTION
**Pull request by str4d**
This currently just wraps a `Column<Fixed>`, but enables us to start writing circuits that can later have their selector usage optimised.

Part of #116.

### CODE REVIEW

File: src/plonk/circuit.rs [hunk 1]
```diff
@@ -126,6 +129,71 @@ impl TryFrom<Column<Any>> for Column<Instance> {
...
+///
+/// struct Config {
+///     a: Column<Advice>,
+///     b: Column<Advice>,
+///     s: Selector,
+/// }
+///
+/// fn circuit_logic<C: Chip>(mut layouter: impl Layouter<C>) -> Result<(), Error> {
+///     let config = layouter.config().clone();
+///     # let config: Config = todo!();
```
Comment (str4d) [inline src/plonk/circuit.rs:171]: This is a trick to make the documentation look correct, while also type-checking correctly in CI.

File: src/plonk/circuit.rs [hunk 2]
```diff
@@ -126,6 +129,71 @@ impl TryFrom<Column<Any>> for Column<Instance> {
...
+impl Selector {
+    /// Enable this selector at the given offset within the given region.
+    pub fn enable<C: Chip>(&self, region: &mut Region<C>, offset: usize) -> Result<(), Error> {
+        // TODO: Ensure that the default for a selector's cells is always zero, if we
+        // alter the proving system to change the global default.
+        // TODO: Add Region::enable_selector method to allow the layouter to control the
+        // selector's assignment.
+        // https://github.com/zcash/halo2/issues/116
+        region
+            .assign_fixed(|| "", self.0, offset, || Ok(C::Field::one()))
```
Comment (str4d) [inline src/plonk/circuit.rs:192]: TODO: How should we annotate selectors being enabled? Should `Selector::enable` take an annotation, or should we compute this automatically (perhaps from a name we provide to `ConstraintSystem::selector`)?

File: src/plonk/circuit.rs [hunk 3]
```diff
@@ -614,6 +687,13 @@ impl<F: Field> ConstraintSystem<F> {
         self.gates.push((name, poly));
     }
 
+    /// Allocate a new selector.
+    pub fn selector(&mut self) -> Selector {
```
Comment (str4d) [inline src/plonk/circuit.rs:691]: We might also want to add `Selector::new(&mut ConstraintSystem)` to match `Permutation::new` as in #199.

File: src/plonk/circuit.rs [hunk 4]
```diff
@@ -479,6 +547,11 @@ impl<F: Field> ConstraintSystem<F> {
         index
     }
 
+    /// Query a selector at a relative position.
+    pub fn query_selector(&mut self, selector: Selector, at: Rotation) -> Expression<F> {
```
Comment (str4d) [inline src/plonk/circuit.rs:551]: If this querying logic was moved onto a gate-specific wrapper around `ConstraintSystem` (similar to how a `Region` will at some point wrap a `Layouter`), we could collect metadata about which columns / selectors are used in which gates (probably more easily than extracting that information from the gate expressions).

**Review by ebfull — APPROVED** [review]

**Comment by str4d** [comment]
Rebased to fix conflicts with #199.

**Pull request merged**